Additional operators

EmielK opened the pull request.

EmielK committed:
Implement perturbed_greedy_insert

N-Wouda commented on heuristic/repair_operators/perturbed_greedy_insert.py:
```diff
@@ -0,0 +1,43 @@
+                heapq.heappush(feasible_routes, (cost, insert_idx, route))
+
+        if len(feasible_routes) != 0:
+            y = min(3, len(feasible_routes))
```
Make `3` into a constant (in constants.py)

N-Wouda replied:
Also: this is `offset` or something. I'd not call it `y` here.

N-Wouda commented on heuristic/repair_operators/perturbed_greedy_insert.py:
```diff
@@ -0,0 +1,43 @@
+        if len(feasible_routes) != 0:
+            y = min(3, len(feasible_routes))
+            nsmallest = heapq.nsmallest(y, feasible_routes)
+            rnd_state.shuffle(nsmallest)
```
Instead do `cost, insert_idx, route = rnd_state.choice(nsmallest)`. It's both faster and easier to read.

N-Wouda commented on heuristic/repair_operators/perturbed_greedy_insert.py:
```diff
@@ -0,0 +1,43 @@
+
+    Perturbed sequential best insertion in Hornstra et al. (2020).
+    """
+    rnd_state.shuffle(current.unassigned)
```
Most of this can be merged with `greedy_insert`, using `y=1` for the usual greedy operator. As to how: look into `_worst` for the destroy operators, and how that's used.

EmielK committed:
Implement cross_route destroy operator

N-Wouda commented:
@EmielK also: this does not work (yet). Travis complains!

EmielK committed:
Add _yth_best_insert

N-Wouda commented on heuristic/repair_operators/_yth_best_insert.py:
```diff
@@ -0,0 +1,41 @@
+            offset = min(min_offset, len(feasible_routes))
+            nsmallest = heapq.nsmallest(offset, feasible_routes)
+
+            cost, insert_idx, route = nsmallest[rnd_state.choice(offset)]
```
Why? Does `rnd_state.choice(nsmallest)` not work?

EmielK replied:
No, I cant get choice to work for things that are not simple lists or ints.  (This is a list of tuples). ``rnd_state.choice(nsmallest)`` gives the following error:
`` 
line 29, in _yth_best_insert
    cost, insert_idx, route = rnd_state.choice(nsmallest)
  File "mtrand.pyx", line 898, in numpy.random.mtrand.RandomState.choice
ValueError: a must be 1-dimensional
``
Im unsure how to get this to work for a list of tuples, or for a random value from a set.

N-Wouda commented on heuristic/repair_operators/_yth_best_insert.py:
```diff
@@ -0,0 +1,41 @@
+from heuristic.functions import create_single_customer_route
+
+
+def _yth_best_insert(min_offset: int, current: Solution,
```
I'd call this `_random_near_best_greedy_insert` I think. That's much more descriptive, and avoids using 'y', which has no meaning in our code. Do _not_ use 'y', it's completely meaningless.

Then `min_offset` can become `max_offset`, with an explanation of how much worse it could potentially be (at most max_offset distance from the optimal insertion point).

EmielK committed:
Fix cross_route

EmielK committed:
Rewrite cross route, minor changes near_best_greedy_insert

Find route is used less now. The entire selection is now removed every time until the customers removed exceeds the required amount rather than one by one until exactly enough customers have been removed.

N-Wouda commented on heuristic/destroy_operators/cross_route.py:
```diff
@@ -0,0 +1,64 @@
+
+    for customer in route.customers[max(idx - 1, 0):
+                                    min(idx + 2, len(route.customers))]:
+        selected.append(customer)
```
Use extend, as in:
```python
selected.extend(route.customers[max(idx...), min(idx...)]
```

N-Wouda replied:
Or even `selected = route.customers[max(idx...), min(idx..)]`

N-Wouda replied:
Or even
```python
idx = ...
return route.customers[max(idx...), min(idx...)]
```
There's no point in making it more complex than it needs to be!

N-Wouda commented on heuristic/destroy_operators/cross_route.py:
```diff
@@ -0,0 +1,64 @@
+    return selected
+
+
+def _remove(route: Route, removed_set: set, customers: set, selected: list):
```
`Set` and `List` - these are types! (import them from `typing`)

N-Wouda commented on heuristic/repair_operators/perturbed_greedy_insert.py:
```diff
@@ -0,0 +1,17 @@
+
+    Perturbed sequential best insertion in Hornstra et al. (2020).
+    """
+
```
Weird space

N-Wouda commented on heuristic/repair_operators/_near_best_greedy_insert.py:
```diff
@@ -0,0 +1,43 @@
+            nsmallest = heapq.nsmallest(offset, feasible_routes)
+
+            cost, insert_idx, route = nsmallest[rnd_state.choice(offset)]
+
```
Weird space

N-Wouda commented on heuristic/repair_operators/_near_best_greedy_insert.py:
```diff
@@ -0,0 +1,43 @@
+                heapq.heappush(feasible_routes, (cost, insert_idx, route))
+
+        if len(feasible_routes) != 0:
+            offset = min(max_offset, len(feasible_routes))
```
`num_smallest` - offset doesn't really mean anything.

N-Wouda commented on heuristic/destroy_operators/cross_route.py:
```diff
@@ -0,0 +1,64 @@
+        selected_1 = \
+            _select_candidate_and_neighbours(route_candidate, candidate)
+        route_other = []
+        selected_2 = []
```
Ugly and does not conform to types. Why not remove candidate and neighbours here, and then remove other and neighbours later on? That cleans up the flow nicely.

N-Wouda commented on heuristic/constants.py:
```diff
@@ -7,3 +7,4 @@
 DECAY = 0.6
 ITERATIONS = 1000
 CRITERION = RecordToRecordTravel(200, 1, step=199 / ITERATIONS)
+MIN_OFFSET = 3
```
Weird name

N-Wouda committed:
Merge branch 'master' into additional-operators

N-Wouda committed:
Fix some small stuff

N-Wouda committed:
Clean-up cross_route, documentation

N-Wouda committed:
Reset main

N-Wouda committed:
Fix constants

N-Wouda committed:
Split Travis and our own configurations

N-Wouda commented:
I think this is OK now @EmielK. Once Travis completes I'll merge this in.

N-Wouda committed:
Fix missing constants
